src/aspire/source/__init__.py

```diff
@@ -370,6 +375,91 @@ def whiten(self, noise_filter):
+        of the noise, and making sure that the mean of the molecule is larger.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending a `Multiple` filter to the generation pipeline.
+
+        :param batch_size: Batch size of images to query.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+
+        logger.info('Apply contrast inversion on source object')
+        L = self.L
+        grid = grid_2d(L, shifted=True)
+        # Get mask indices of signal and noise samples assuming molecule
+        signal_mask = (grid['r'] < 0.5)
+        noise_mask = (grid['r'] > 0.8)
+
+        # Calculate mean values in batch_size
+        signal_mean = 0.0
+        noise_mean = 0.0
+
+        for i in range(0, self.n, batch_size):
+            images = self.images(i, batch_size).asnumpy()
+            signal = images * np.expand_dims(signal_mask, 2)
+            noise = images * np.expand_dims(noise_mask, 2)
+            signal_mean += np.sum(signal)
+            noise_mean += np.sum(noise)
+        signal_denominator = self.n * np.sum(signal_mask)
+        noise_denominator = self.n * np.sum(noise_mask)
+        signal_mean /= signal_denominator
+        noise_mean /= noise_denominator
+
+        if signal_mean < noise_mean:
+            logger.info('Need to invert contrast')
+            scale_factor = -1.0 * np.ones(self.n)
+        else:
+            logger.info('No need to invert contrast')
+            scale_factor = 1.0 * np.ones(self.n)
+
+        logger.info('Adding Scaling Xform to end of generation pipeline')
+        self.generation_pipeline.add_xform(Multiply(scale_factor))
+
+        # Invalidate images
+        self._im = None
+
+    def normalize_background(self, bg_radius=1.0, do_ramp=True):
+        """
+        Normalize the images by the noise background
+
+        This is done by shifting the image density by the mean value of background
+        and scaling the image density by the standard deviation of background.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending the `Add` and `Multiple` filters to the generation pipeline.
+
+        :param bg_radius: Radius cutoff to be considered as background (in image size)
+        :param do_ramp: When it is `True`, fit a ramping background to the data
+            and subtract. Namely perform normalization based on values from each image.
+            Otherwise, a constant background level from all images is used.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+
+        logger.info(f'Normalize background on source object with radius '
+                    f'size of {bg_radius} and do_ramp of {do_ramp}')
+        self.generation_pipeline.add_xform(
+            LambdaXform(normalize_bg, bg_radius=bg_radius, do_ramp=do_ramp))
+
+        # Invalidate images
```

Comment:
This will have to be removed. If #237 is merged first, we rebase on top of `develop` and remove if. If this PR is merged first, we rebase #237 and remove the line there.

Reply:
PR #237 was merged first. I have rebased this preprocess branch based on the latest develop branch. The statements of invalidating images has been removed in commit  4c3334e.